### PR TITLE
Fix scipy import failure #2338

### PR DESCRIPTION
**Pull request by hoodmane**
Trying to understand this bug made me realize that I am very deeply confused about a lot of stuff. But the actual cause of it is very simple: I patched this into `__init__.py`:
```py
if "pyodide_js" in sys.modules:
   from pyodide_js._module import loadDynamicLibrary
   loadDynamicLibrary('/lib/python3.10/site-packages/scipy/linalg/_flapack.so')
```
I uh... I have no idea what the conditional is supposed to mean. As far as I can tell, `if "pyodide_js" in sys.modules:` is just total nonsense. But it does explain the strange behavior of the bug. Anyways, I added a new patch with a detailed explanation.

I still don't really understand why importing scipy ever works. But this is okay, if the tests pass then who cares why it works right??

### CODE REVIEW

**Comment by rth** [comment]
Thanks for the patch!

I don't understand why we need to load it explicitly as opposed to during import. I get there is some Fortran in there, but what makes it different from all the other .so files which as loaded with the global mechanism? Also normally `_flapack.so` is only loaded when importing `scipy.linlalg.lapack` and not the root scipy import.

**Comment by hoodmane** [comment]
> Also normally `_flapack.so` is only loaded when importing `scipy.linlalg.lapack` and not the root scipy import.

Fair enough I moved the patch there.

**Comment by hoodmane** [comment]
I believe there are three ways that symbols are loaded from shared libraries (I think this is true in both linux and emscripten):

1. `blah.so` is always loaded before `some.so`, and the loader looks up the symbols in `blah.so` at load time and writes the memory addresses of the `some.so` symbols directly into `blah.so`. On Wasm this is a bit different, but the loader locates those symbols in `blah.so` and puts them into the imports for the module as part of linking.
2. `some.so` file and `blah.so` are loaded in any order, `some.so` lazily loads the symbols it needs from the Global Object Table (GOT).
3. `some.so` uses `dlopen('blah.so')` and `dlsym`.

#2 is the only option that breaks cycles. Generally #1 and #2 are done automatically by the compiler, #3 requires breaking cycles. Different `.so` files can have symbols with the same name as long as they don't go into the GOT, only one symbol of each name can go in the GOT. 

Python loads packages using dlsym to load `PyInit_package_name` and then calls `PyInit_package_name` which inserts function pointers into a bunch of tables (effectively it using a fourth unrelated loaded system custom designed in Python).


Anyways, `loadDynamicLibrary` is somewhat poorly named. It doesn't load the library. What it does is it dumps the symbols from an already loaded library into the GOT. Scipy tries to load symbols from LAPACK via the GOT so we need LAPACK symbols to be in there. We have stuffed extra symbols from future versions of LAPACK into `_flapack.so`, and other Scipy libs try to look them up in the GOT so we need to get them in there.

**Comment by ryanking13** [comment]
Thanks for working on this @hoodmane!

> `loadDynamicLibrary` is somewhat poorly named. It doesn't load the library. What it does is it dumps the symbols from an already loaded library into the GOT.

Huh, this is interesting...

File: packages/scipy/patches/0001-loadDynamicLibrary-flapack.patch [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
++++ b/scipy/linalg/lapack.py
+@@ -809,6 +809,15 @@ All functions
+    ilaver
+ 
+ """
++
++import sysconfig
++import pyodide_js
++ext_suffix = sysconfig.get_config_var("EXT_SUFFIX")
++pyodide_js._module.loadDynamicLibrary(f'/lib/python3.10/site-packages/scipy/linalg/_flapack{ext_suffix}')
```
Comment (ryanking13) [inline packages/scipy/patches/0001-loadDynamicLibrary-flapack.patch:44]: For the future update:

```suggestion
+pyodide_js._module.loadDynamicLibrary(f'/lib/python{sys.version_info.major}.{sys.version_info.minor}/site-packages/scipy/linalg/_flapack{ext_suffix}')
```

Comment (hoodmane) [reply packages/scipy/patches/0001-loadDynamicLibrary-flapack.patch:44]: Or maybe we can use:
```py
from site import getsitepackages
f"{getsitepackages}[0]}/scipy/linalg/_flapack{ext_suffix}"
```

Comment (ryanking13) [reply packages/scipy/patches/0001-loadDynamicLibrary-flapack.patch:44]: Yes, that looks better.

**Pull request merged**